### PR TITLE
feat: shared cache for manual peering

### PR DESCRIPTION
**Pull request by hacdias**
- [x] Depends on the folllowing. Please review and merge those before this one. I will rebase once those are merged.
    - [x] #113 
    - [x] https://github.com/ipfs/boxo/pull/597

CC #109  #27

### CODE REVIEW

File: setup.go [hunk 1]
```diff
@@ -319,17 +322,39 @@ func Setup(ctx context.Context, cfg Config, key crypto.PrivKey, dnsCache *cached
...
 		// ProviderSearchDelay: default is 1 second.
-		bsclient.ProviderSearchDelay(time.Second),
-		bsclient.WithoutDuplicatedBlockStats(),
+		bitswap.ProviderSearchDelay(time.Second),
+		bitswap.WithoutDuplicatedBlockStats(),
+
+		// ---- Server Options
+		bitswap.WithPeerBlockRequestFilter(pbrf),
+		bitswap.ProvideEnabled(true),
 	)
```
Comment (hacdias) [inline setup.go:358]: Another option here would be to create only a client with `bsclient.New` if peering cache is disabled. But to me it seemed that a peer block request filter that always returns `false` should do the same thing. Depends a bit more on the internals of the bitswap package.

File: setup.go [hunk 2]
```diff
@@ -319,17 +322,44 @@ func Setup(ctx context.Context, cfg Config, key crypto.PrivKey, dnsCache *cached
...
 		// ProviderSearchDelay: default is 1 second.
-		bsclient.ProviderSearchDelay(time.Second),
-		bsclient.WithoutDuplicatedBlockStats(),
+		bitswap.ProviderSearchDelay(time.Second),
+		bitswap.WithoutDuplicatedBlockStats(),
+
+		// ---- Server Options
+		bitswap.WithPeerBlockRequestFilter(peerBlockRequestFilter),
+		bitswap.ProvideEnabled(provideEnabled),
 	)
```
Comment (hacdias) [inline setup.go:363]: Another option here would be to create only a client with `bsclient.New` if peering cache is disabled. But to me it seemed that a peer block request filter that always returns `false` should do the same thing. Depends a bit more on the internals of the bitswap package.

File: setup_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		{0, 2},
+		{0, 1},
+	}, false)
+}
+
+func TestPeeringCache(t *testing.T) {
+	nodes := mustPeeredNodes(t, [][]int{
+		{1},
+		{0},
+		{},
```
Comment (lidel) [inline setup_test.go:125]:
```suggestion
func TestPeeringCache(t *testing.T) {
	nodes := mustPeeredNodes(t, [][]int{
		{1}, // 0 peered to 1
		{0}, // 1 peered to 0
		{},  // 2 not peered to anyone
```

File: setup_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		} else {
+			require.Error(t, err)
+		}
+	}
+
+	err := nodes[0].bsrv.AddBlock(ctx, bl)
+	require.NoError(t, err)
+
+	checkBitswap(1, true)
+	checkBitswap(2, false)
```
Comment (lidel) [inline setup_test.go:149]:
```suggestion
  // confirm peering enables cache sharing, and bitswap retrieval from safelisted node works
	checkBitswap(1, true)
	// confirm bitswap providing is disabled by default (no peering)
	checkBitswap(2, false)
```

File: main.go
Comment (lidel) [inline main.go:1]: @hacdias mind adding extra code when `--seed` is present, to calculate peerids from placeholders passed to `--peering`? (the idea from https://github.com/ipfs/rainbow/pull/111#discussion_r1554174855)

To simplify infra scripts we would pass `--peering=/dns/example/tcp/4001/p2p/rainbow-seed/42` and rainbow would replace that on the fly with actual peerid `/dns/example/tcp/4001/p2p/12..`


Good place for search and replace of `/p2p/rainbow-seed/(\d+)` suffix:

https://github.com/ipfs/rainbow/blob/19723fe3c522dba0daa861bf64f02dad30fde7e2/main.go#L280-L281

**Review by lidel — CHANGES_REQUESTED** [review]

File: setup.go (hunk 2 above)
Comment (aschmahmann) [reply setup.go:363]: This might be ok, but we should measure our libp2p egress bandwidth to hosts we don't control to see to validate that this is true.

**Review by lidel — APPROVED** [review]
lgtm, pushed small cosmetics, should be good to ship as v1.1.0

**Pull request merged**